File: bindings/R/fm/src/color.cpp
#include "color.h"


#include <Rcpp.h>


void fmColor_show(const fort::myrmidon::Color * c) {
	Rcpp::Rcout << "fmColor ( R = " << (int)std::get<0>(*c)
	            << ", G = " << (int)std::get<1>(*c)
	            << ", B = " << (int)std::get<2>(*c)
	            << ")\n";
}

int fmColor_red(const fort::myrmidon::Color * c) {
	return std::get<0>(*c);
}

int fmColor_green(const fort::myrmidon::Color * c) {
	return std::get<1>(*c);
}

int fmColor_blue(const fort::myrmidon::Color * c) {
	return std::get<2>(*c);
}

fort::myrmidon::Color * fmRGBColor(SEXP xx) {
	Rcpp::IntegerVector v(xx);
	if ( v.size() != 3 ) {
		throw std::runtime_error("fmRGBColor needs an integer vector of size 3");
	}
	return new fort::myrmidon::Color(std::clamp(v[0],0,255),
	                                 std::clamp(v[1],0,255),
	                                 std::clamp(v[2],0,255));
}

fort::myrmidon::Color * fmDefaultPaletteColor(size_t i) {
	return new fort::myrmidon::Color(fort::myrmidon::Palette::Default().At(i));
}

RCPP_MODULE(color) {
	Rcpp::class_<fort::myrmidon::Color>("fmColor")
		.const_method("show",&fmColor_show)
		.const_method("red",&fmColor_red)
		.const_method("green",&fmColor_green)
		.const_method("blue",&fmColor_blue)
		;

	Rcpp::function("fmRGBColor",&fmRGBColor);
	Rcpp::function("fmDefaultPaletteColor",&fmDefaultPaletteColor);



}

File: bindings/R/FortMyrmidon/R/Color.R
#' @name fmColor
#' @title Reference class for a RGB color
#' @description this is a wrapper for the C++ class
#'     \href{https://formicidae-tracker.github.io/studio/docs/latest/api/#fort::myrmidon/fort::myrmidon::Color}{fort::myrmidon::Color}. Since
#'     \code{std::tuple} does not exist in R, it defines their own
#'     custom methods.
#'     \cr\cr
#'     fmColor can be created with \code{\link{fmRGBColor}} or
#'     \code{\link{fmDefaultPaletteColor}}
#'
#' @field red the red component of the color as an integer between 0 and 255
#' @field green the green component of the color as an integer between 0 and 255
#' @field blue the blue component of the color as an integer between 0 and 255
NULL

#' @name fmRGBColor
#' @title Creates an \code{\link{fmColor}} from its RGB value
#' @param xx an integer vectorr of size 3 with the RGB value in this
#'     order
#' @return the wanted \code{\link{fmColor}}
#' @examples
#' white <- fmRGBColor(c(255,255,255))
#' pureBlue <- fmRGBColor(c(0,0,255))
NULL

#' @name fmDefaultPaletteColor
#' @title Creates an \code{\link{fmColor}} from myrmidon palette
#' @param i an integer to draw a color from
#' @description myrmidon defines a color-blind friendly palette of 7
#'     colors that can be accessed with this method. #i does not need
#'     to be between 0 and 6 color will cycle through the 7 available
#'     colors.
#' @return the wanted \code{\link{fmColor}}
#' @examples
#' # the two fmColor below will be distinguishable for most color-blind people
#' firstColor <- fmDefaultPaletteColor(0)
#' secondColor <- fmDefaultPaletteColor(1)
NULL
